edgelet/hsm-sys/azure-iot-hsm-c/src/edge_hsm_client_x509.c

```diff
@@ -1,72 +1,501 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <math.h>
+#include <stdbool.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <stdbool.h>
+#include <time.h>
 
 #include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/crt_abstractions.h"
 #include "hsm_client_data.h"
 #include "hsm_log.h"
+#include "hsm_utils.h"
+
+extern const char* const EDGE_DEVICE_ALIAS;
+extern const char* const ENV_DEVICE_CERTIFICATE_PATH;
+extern const char* const ENV_DEVICE_PRIVATE_KEY_PATH;
+extern const char* const ENV_REGISTRATION_ID;
+
+static bool g_is_x509_initialized = false;
+static unsigned int g_ref_cnt = 0;
 
 int hsm_client_x509_init()
 {
-    return 0;
+    int result;
+
+    if (!g_is_x509_initialized)
+    {
+        result = hsm_client_crypto_init();
+        if (result == 0)
+        {
+            g_is_x509_initialized = true;
+        }
+    }
+    else
+    {
+        result = 0;
+    }
+
+    return result;
 }
 
 void hsm_client_x509_deinit()
 {
+    if (!g_is_x509_initialized)
+    {
+        LOG_ERROR("hsm_client_x509_init not called");
+    }
+    else
+    {
+        if (g_ref_cnt == 0)
+        {
+            g_is_x509_initialized = false;
+            hsm_client_crypto_deinit();
+        }
+    }
 }
 
-void iothub_hsm_free_buffer(void * buffer)
+void edge_x509_hsm_free_buffer(void * buffer)
 {
     if (buffer != NULL)
     {
         free(buffer);
     }
 }
 
-HSM_CLIENT_HANDLE iothub_x509_hsm_create()
+HSM_CLIENT_HANDLE edge_x598_hsm_create()
 {
-    LOG_ERROR("API unsupported");
-    return (HSM_CLIENT_HANDLE)NULL;
+    HSM_CLIENT_HANDLE result;
+
+    if (!g_is_x509_initialized)
+    {
+        LOG_ERROR("hsm_client_x509_init not called");
+        result = NULL;
+    }
+    else
+    {
+        const HSM_CLIENT_CRYPTO_INTERFACE* interface = hsm_client_crypto_interface();
+        if (interface == NULL)
+        {
+            LOG_ERROR("hsm_client_crypto_interface returned NULL");
+            result = NULL;
+        }
+        else
+        {
+            result = interface->hsm_client_crypto_create();
+            if (result != NULL)
+            {
+                g_ref_cnt++;
```

Comment:
Who does the locking for these ref counts?

Reply:
All the HSM APIs expect the application to perform the synchronization. We do this in the Edglet via the ARC mutex owned by iotedged.